src/QsCompiler/QirGeneration/DebugInfoManager.cs

```diff
@@ -0,0 +1,575 @@
+            if (subProgram != null && dIType != null)
+            {
+                if (dIType.Name.Equals(DebugTypeNotSupportedMessage))
+                {
+                    name += " [value display for this type is not currently supported]";
+                }
+
+                DebugPosition debugLocation = ConvertToDebugPosition(absolutePosition);
+
+                // create the debug info for the local variable
+                DILocalVariable dIVar = dIBuilder.CreateLocalVariable(
+                    subProgram,
+                    name,
+                    dIBuilder.CompileUnit?.File,
+                    debugLocation.Line,
+                    dIType,
+                    alwaysPreserve: true,
+                    DebugInfoFlags.None);
+
+                DILocation dILocation = new DILocation(
+                    this.Context,
+                    debugLocation.Line,
+                    debugLocation.Column,
+                    subProgram);
+
+                if (this.sharedState.CurrentBlock != null)
+                {
+                    if (isMutableBinding)
+                    {
+                        // variable is represented as a pointer to the value
+                        Value variable = ((PointerValue)value).Pointer;
+                        dIBuilder.InsertDeclare(
+                        storage: variable,
+                        varInfo: dIVar,
+                        location: dILocation,
+                        insertAtEnd: this.sharedState.CurrentBlock);
+                    }
+                    else
+                    {
+                        // variable is represented as the value itself
+                        Value variable = value.Value;
+                        dIBuilder.InsertValue(
+                            value: variable,
+                            varInfo: dIVar,
+                            location: dILocation,
+                            insertAtEnd: this.sharedState.CurrentBlock);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Creates a global function. If the debug flag is set to true, includes debug information
+        /// with the global function.
+        /// </summary>
+        /// <param name="spec">The <see cref="QsSpecialization"/> for the function.</param>
+        /// <param name="mangledName">The mangled name for the function.</param>
+        /// <param name="signature">The signature for the function.</param>
+        /// <returns>The <see cref="IrFunction"/> corresponding to the given Q# type.</returns>
+        internal IrFunction CreateGlobalFunction(QsSpecialization spec, string mangledName, IFunctionType signature)
+        {
+            if (!Config.DebugSymbolsEnabled)
+            {
+                return this.CreateFunctionNoDebug(mangledName, signature);
+            }
+
+            if (spec.Kind == QsSpecializationKind.QsBody)
```

Comment:
For adjoint and controlled operations, this just creates a function with no debug info but it is not clear in the code that it is doing that. I would suggest encapsulating the contents of this `if` statement in a method and using a `switch` statement here to make clear that `QsSpecializationKind.QsBody` is the only kind for which debug symbols are currently supported and that for the other ones debug symbols are explicitly not supported:
```csharp
var func = spec.Kind switch
    {
        QsSpecializationKind.QsAdjoint => this.CreateFunctionNoDebug(mangledName, signature),
        QsSpecializationKind.QsBody => this.CreateQsBodyFunction(spec, mangledName, signature),
        QsSpecializationKind.QsControlled => this.CreateFunctionNoDebug(mangledName, signature),
        QsSpecializationKind.QsControlledAdjoint => this.CreateFunctionNoDebug(mangledName, signature),
        _ => throw new ArgumentException($"Unsupported QsSpecializationKind '{spec.Kind}'"),
    }
```

Reply:
I couldn't use the switch statement like this because it expects a constant instead of something like QsSpecializationKind.QsAdjoint. There's probably a workaround I could figure out, but I did something else to make it clearer. Could you let me know if this fix is what you were looking for?

Reply:
That's interesting, `QsSpecializationKind.QsBody` is a constant and it is used in a switch statement [here](https://github.com/microsoft/qsharp-compiler/blob/a4ef30735d856069f92ef33e00e92adcf433f49b/src/QsCompiler/BondSchemas/CompilerObjectTranslator.cs#L380). I'm ok with the way it currently is but I would prefer the `switch` expression since makes it clear what are all the possible specialization kinds.